Added recipe for libsdl3

captain0xff opened the pull request:
I have added the initial recipe for libsdl3. The recipe is basically taken from the SDL2 recipe. Things related to SDL2main, sdl2-config and SDL2-2.0 are gone since they are not relevant for SDL3. SDL_main is now a header only library and there is no sdl3-config as now pkg-config provides everything (I hope a fixPkgconfig suffices for that).

Known issues:
1. ~~I have only tested in a VM but keyboard related events don't seem to be registered (I will look into this).~~

I have run the tests to check for issues but I think they don't cover all the subsystems well so some more thorough testing of the various subsystems is needed.

Begasus commented:
You didn't create a branch for this, although it's not a required it's adviced to not run into conflicts.

Begasus commented:
Test case should be enabled (at least on the first run) to be able to compare on future updates. At the moment "96% tests passed, 1 tests failed out of 23", the one failing crashes also (filesystem)

Begasus commented on media-libs/libsdl3/libsdl3-3.2.4.recipe:
```diff
@@ -0,0 +1,74 @@
+
+PROVIDES="
+    libsdl3$secondaryArchSuffix = $portVersion compat >= 3.0
+    lib:libsdl3$secondaryArchSuffix = 0.200.4 compat >= 0
```
Use libVersionCompat for this, also version is wrong here:
```
-- Installing: /packages/libsdl3-3.2.4-1/.self/lib/libSDL3.so.0.2.4
-- Installing: /packages/libsdl3-3.2.4-1/.self/lib/libSDL3.so.0
-- Installing: /packages/libsdl3-3.2.4-1/.self/lib/libSDL3.so
```

captain0xff replied:
I have changed the names to uppercase as instructed.
And regarding the other issue, I was not really sure about the versioning part so I just followed the sdl2 port
```
	libsdl2${secondaryArchSuffix}_devel = $portVersion compat >= 2.0
	cmd:sdl2_config$secondaryArchSuffix = $portVersion compat >= 2.0
	devel:libSDL2$secondaryArchSuffix = 0.3000.10 compat >= 0
	devel:libSDL2_2.0$secondaryArchSuffix = 0.3000.10 compat >= 0
	devel:libSDL2main$secondaryArchSuffix
```
It seems to use 0.3000.10 for v2.30.10. I tried checking the build recipe docs but it isn't very clear about this either. Can you kindly guide me here?

captain0xff replied:
Pushing rest of the changes.

Begasus replied:
The output you pasted is the output created by the recipe, not what's actually in there, for libversion you can check the install log near the end of the build and see the output I posted earlier.
You can also double click the created package in `~/haikuports/packages` to see it's content.

captain0xff replied:
Uhm, what I pasted is not any output. It's taken from here: https://github.com/haikuports/haikuports/blob/095aa6c1270671880d1973e4e6931f016863d046/media-libs/libsdl2/libsdl2-2.30.10.recipe#L28

captain0xff replied:
I hope it's correct this time?

Begasus commented on media-libs/libsdl3/libsdl3-3.2.4.recipe:
```diff
@@ -0,0 +1,74 @@
+    ninja -C build install
+
+    rm $libDir/libSDL3_test.a
+    rm $libDir/cmake/SDL3/SDL3testTargets-release.cmake
```
Any reason to remove this file?

Begasus replied:
Just saw that building sdl3_image fails to find the static library, this shouldn't be removed I think and should be moved to the _devel package, then sed for the path with:
```
	sed -i 's,${_IMPORT_PREFIX}/lib',$developDir/lib, \
		$libDir/cmake/SDL3/SDL3testTargets-release.cmake
```

captain0xff replied:
Oh, I see. I saw that the sdl2 port was doing that so just followed that thinking there is some issues it on haiku,

Begasus replied:
Normaly we delete the static libraries (you will get a policy error anyway when libfoo.so and libfoo.a is installed with haikuporter), in this case it's not about libfoo.a but libfoo_test.a and we probably should keep this.

jmairboeck replied:
> ```
> 	sed -i 's,${_IMPORT_PREFIX}/lib',$developDir/lib, \
> 		$libDir/cmake/SDL3/SDL3testTargets-release.cmake
> ```

Does the original code already handle the secondaryArch subdir or should that be `$developLibDir` instead of `$developDir/lib`?

Begasus replied:
My bad on this one :)

captain0xff replied:
Done

captain0xff commented:
> You didn't create a branch for this, although it's not a required it's adviced to not run into conflicts.

I will keep this in mind from next time.

captain0xff commented:
Also, I commented out glu related stuff because I wasn't sure if it's actually needed and the library seemed to fine work without it.

Begasus commented:
> Also, I commented out glu related stuff because I wasn't sure if it's actually needed and the library seemed to fine work without it.

I wouldn't worry about that for now, first thing to do would be to address given comments and get the recipe in shape according to our guidlines: https://github.com/haikuports/haikuports/wiki/HaikuPorter-Guidelines

jmairboeck commented:
> Also, sorry for the generic commit message. I was trying to push from haiku itself and it was not working initially and after trying a few times I put a generic message to avoid typing the same thing again and again. And then it randomly decided to work.

That shouldn't matter too much, as usually the commits from PRs get squashed anyway when merging.

captain0xff commented:
> Test case should be enabled (at least on the first run) to be able to compare on future updates. At the moment `96% tests passed, 1 tests failed out of 23`, the one failing crashes also (testfilesystem)
> 
> ```
> 12/23 Test #12: testfilesystem ...................Signal 21***Exception:   2.70 sec
> base path: '/sources/SDL3-3.2.4/build/test/'
> ```

Should I use the TEST() block for this?

jmairboeck commented on media-libs/libsdl3/libsdl3-3.2.4.recipe:
```diff
@@ -0,0 +1,77 @@
+ARCHITECTURES="all !x86_gcc2"
+SECONDARY_ARCHITECTURES="x86"
+
+libVersion="$portVersion"
```
According to @Begasus in https://github.com/haikuports/haikuports/pull/11770#discussion_r1952040426, this should be `"0.2.4"`.

captain0xff replied:
Sorry, stupid mistake.

jmairboeck commented on media-libs/libsdl3/libsdl3-3.2.4.recipe:
```diff
@@ -0,0 +1,77 @@
+libVersionCompat="$libVersion compat >= ${libVersion%%.*}"
+
+PROVIDES="
+    libsdl3$secondaryArchSuffix = $libVersionCompat
```
The package name PROVIDE must always use `$portVersion`, optionally with some `compat` attached.

```suggestion
    libsdl3$secondaryArchSuffix = $portVersion compat >= 3.0
```

jmairboeck commented on media-libs/libsdl3/libsdl3-3.2.4.recipe:
```diff
@@ -0,0 +1,77 @@
+PROVIDES_devel="
+    libSDL3${secondaryArchSuffix}_devel = $portVersion
+    devel:libSDL3$secondaryArchSuffix = $libVersionCompat
+    devel:libSDL3_test$secondaryArchSuffix = $portVersion
```
does this have the same SO version as libSDL3? Then you can reuse `$libVersionCompat` here, otherwise it should have its own version variable and use that.

captain0xff replied:
Uhm, it's a static library

jmairboeck replied:
I think $portVersion is okay in this case. Sorry for the confusion.

captain0xff replied:
No problem :)

jmairboeck commented on media-libs/libsdl3/libsdl3-3.2.4.recipe:
```diff
@@ -0,0 +1,77 @@
+{
+    ninja -C build install
+
+    sed -i 's,${_IMPORT_PREFIX}/lib',$developDir, \
```
I think this should be 
```suggestion
    sed -i 's,${_IMPORT_PREFIX}/lib',$developLibDir, \
```
but to be sure that should be tested with some other package which actually uses this cmake file.

captain0xff replied:
Understood. I will try with SDL3_image

captain0xff replied:
$developLibDir seems to work

captain0xff commented:
> I have only tested in a VM but keyboard related events don't seem to be registered (I will look into this).

This issue is fixed. Thanks to @icculus for [this](https://github.com/libsdl-org/SDL/commit/1354affd288b8bc3b354a155f27201cda31dfa7f).

captain0xff commented:
So, I read https://github.com/haikuports/haikuports/wiki/HaikuPorter-Guidelines as instructed and I think the recipe is more or less following the guidelines? I have addressed all the comments.
What do I need to do next? Enable the tests? If yes can somebody kindly answer [this](https://github.com/haikuports/haikuports/pull/11770#issuecomment-2654729990).

Begasus commented:
You are still using spaces instead of tab(s) where needed (indents)

Begasus commented on media-libs/libsdl3/libsdl3-3.2.4.recipe:
```diff
@@ -0,0 +1,77 @@
+INSTALL()
+{
+    ninja -C build install
+
```
You can remove the dataDir, it only contains  a renamed Zlib license.

Begasus replied:
Not resolved, $dataDir still around.

Begasus commented:
> > I have only tested in a VM but keyboard related events don't seem to be registered (I will look into this).
> 
> This issue is fixed. Thanks to @icculus for [this](https://github.com/libsdl-org/SDL/commit/1354affd288b8bc3b354a155f27201cda31dfa7f).

This hasn't been included in this release, so maybe you could backport it here?

Begasus commented:
For the test, you have to enable this in BUILD (it's disabled by default) and add a TEST case then.

captain0xff commented:
Done. Addressed all the requested changes. And the patch works on my end. Also, all the tests pass except filesystem like you observed.

Begasus commented on media-libs/libsdl3/libsdl3-3.2.4.recipe:
```diff
@@ -0,0 +1,84 @@
+
+PROVIDES="
+	libsdl3$secondaryArchSuffix = $portVersion compat >= 3.0
+	lib:libsdl3$secondaryArchSuffix = $libVersionCompat
```
lib:libSDL3

Begasus commented on media-libs/libsdl3/libsdl3-3.2.4.recipe:
```diff
@@ -0,0 +1,84 @@
+	"
+
+PROVIDES_devel="
+	libSDL3${secondaryArchSuffix}_devel = $portVersion compat >= 3.0
```
libsdl3${secondaryArchSuffix}_devel

Begasus commented on media-libs/libsdl3/libsdl3-3.2.4.recipe:
```diff
@@ -0,0 +1,84 @@
+	devel:libSDL3_test$secondaryArchSuffix = $portVersion
+	"
+REQUIRES_devel="
+	libSDL3$secondaryArchSuffix == $portVersion base
```
libsdl3$secondaryArchSuffix

Begasus commented on media-libs/libsdl3/libsdl3-3.2.4.recipe:
```diff
@@ -0,0 +1,84 @@
+		$developDir \
+		$libDir/cmake
+
+	rm -rf $prefix/bin
```
Didn't check this change, but I don't think there is anything installed in $prefix/bin?

captain0xff replied:
You right. Ig this was there in the SDL2 recipe because of sdl2-config.

Begasus commented:
Thanks so far, still a few cosmetic changes, will give it another spin here.

captain0xff commented:
Sorry, I think I am misunderstanding something here because I am unable to find anything related to dataDir in the recipe. And the there is no such thing in the review line. Here is what I am seeing.
![image](https://github.com/user-attachments/assets/334cc827-6581-42f7-bddb-25d9a0e3ed54)

Addressed rest of the reviews.

Begasus commented on media-libs/libsdl3/libsdl3-3.2.4.recipe:
```diff
@@ -0,0 +1,77 @@
+    fixPkgconfig
+
+    packageEntries devel \
+        $developDir $dataDir $libDir/cmake
```
That is the problem if users mark things as resolved when they are not, @captain0xff here it was mentioned.

captain0xff commented:
@Begasus done. Removed the dataDir. Sorry, for the confusion. Actually, what happened was that the review just below the one where you asked me to remove the dataDir also mentioned about dropping the dataDir. So, I thought that you were asking me to remove all references of dataDir from the recipe itself lol. I apologize again for the misunderstanding.

captain0xff commented:
Fixing the ling problem in a moment.

Begasus commented:
Thanks for sitting this one out, it can be scary on the first recipes, but it worked out well.
Let's see where this leads us in the future. 👍

captain0xff commented:
Welcome. And thank you as well for your patience in guiding me through the process!
